Assignment 4/src/FractionalWrapper.sol

```diff
@@ -0,0 +1,302 @@
+    uint256 constant fractionDenominator = 10**27;
+
+    /// @notice The total amount of assets managed by the vault.
+    /// @dev This simply calls the balance of this address. This means that not
+    ///     all assets may be redeemable by an account: we could have received
+    ///     a generous gift. These would still add up to the total.
+    /// @return totalManagedAssets The assets managed by this contract.
+    function totalAssets()
+        external
+        view
+        override
+        returns (uint256 totalManagedAssets)
+    {
+        return token.balanceOf(address(this));
+    }
+
+    /// @notice Create a new wrapper.
+    /// @param _token The token to use as collateral.
+    /// @param _fraction The fraction at which to mint/withdraw the tokens.
+    /// @dev See the caveats on `fraction` for how to use the `_fraction`
+    ///     parameter.
+    constructor(IERC20 _token, uint256 _fraction) {
+        token = _token;
+        fraction = _fraction;
+    }
+
+    /// @notice Update the fraction at which shares can be minted/burned. This
+    ///     function can only be called by the current owner of the contract.
+    /// @param _fraction The new fraction at which shares can be minted and
+    ///     burned.
+    /// @dev See the caveats on `fraction` for how to use the `_fraction`
+    ///     parameter.
+    function setFraction(uint256 _fraction) external onlyOwner {
+        fraction = _fraction;
+    }
+
+    /// @notice Compute how many shares have the value of a given amount of assets.
+    /// @param assets The amount of assets to do the calculation with.
+    /// @return shares The amount of shares, denoted a precision of `decimals`.
+    function convertToShares(uint256 assets)
+        public
+        view
+        override
+        returns (uint256 shares)
+    {
+        return (assets * fraction) / fractionDenominator;
+    }
+
+    /// @notice Compute how many assets have the value of a given amount of shares.
+    /// @param shares The shares to do the computation on.
+    /// @return assets How many assets are worth the provided `shares`.
+    function convertToAssets(uint256 shares)
+        public
+        view
+        override
+        returns (uint256 assets)
+    {
+        return (shares * fractionDenominator) / fraction;
+    }
+
+    /// @notice The maximal deposit.
+    /// @dev There is no maximal deposit.
+    function maxDeposit(address)
+        external
+        pure
+        override
+        returns (uint256 maxAssets)
+    {
+        return type(uint256).max;
+    }
+
+    /// @notice Compute how many shares would be obtained by depositing in the
+    ///     contract.
+    /// @param assets The amount of assets to be deposited.
+    /// @dev For this contract, this is is the same as `convertToShares`.
+    function previewDeposit(uint256 assets)
+        external
+        view
+        override
+        returns (uint256 shares)
+    {
+        return convertToShares(assets);
```

Comment:
Since `convertToShares` is a public function, calling it here is a CALL and costs 2000+ gas.

You need a public `convertToShares` function to conform to EIP4626. That one can call an internal `_convertToShares` function that has the actual logic. Internal calls are a JUMP can cost something like 10 gas. Then `previewDeposit` can call `_convertToShares` and avoid the public CALL. Same elsewhere.